specification/sections/06-tracking.md

```diff
@@ -10,35 +10,88 @@ toc_max_heading_level: 4
 
 ## Overview
 
-Experimentation is a primary use case for feature flags.
-In practice, this often means flag variants are assigned to users at random or in accordance with a business rule, while the impact of the assigned variant on some business objective is measured.
-Vendors and custom solutions often support a _tracking_ or _goal measuring_ API to facilitate the measurement of these business objectives.
-
-### Goals
-
-- Develop official terminology to support consistent implementation
-- Specify a flexible API widely compatible with basic vendor functionality
-  - Define tracking event payload
-  - Define tracking event identifier
-  - Support A/B testing and experimentation use-cases
-  - Support client and server paradigms
-  - Provide recommendations around: 
-    - Async vs sync
-    - Flushing mechanisms
-    - Event batching
-
-### Non-goals
-
-- Creating an experimentation platform
-- Covering every user-tracking use case
-  - We will not define any data aggregation mechanisms
-  - We will not focus on "metrics", but instead, "facts"
-
-### Design Principles
-
-We value the following:
-
-- Adherence to, and compatibility with OpenFeature semantics
-- Maximum compatibility and ease-of-adoption for existing solutions
-- Minimum traffic and payload size
-- Ease-of-use for application authors, integrators, and provider authors (in that order)
+The `tracking API` enables the association of feature flag evaluations with subsequent actions or application states, in order to facilitate experimentation, and analysis of the impact of feature flags on business objectives.
+
+Combined with hooks which report feature flag evaluations to the analytics platform in question, tracking can allow for robust experimentation even for flag management systems that don't support tracking directly.
+
+```mermaid
+sequenceDiagram
+  Evaluation API->>+Tracking Hook: evaluate
+  Tracking Hook->>Analytics Platform: before
+  Tracking Hook->>Analytics Platform: after
+  Tracking Hook->>-Evaluation API: evaluate
+  Evaluation API->>Analytics Platform: track
+```
+
+### 6.1. Tracking API
+
+#### Condition 6.1.1
+
+> The implementation uses the dynamic-context paradigm.
+
+see: [dynamic-context paradigm](../glossary.md#dynamic-context-paradigm)
+
+##### Conditional Requirement 6.1.1.1
+
+> The `client` **MUST** define a function for tracking the occurrence of a particular action or application state, with parameters `occurrence key` (string, required), `evaluation context` (optional) and `occurrence details` (optional) which returns nothing.
+
+```java
+// example tracking occurrence recording that a subject reached a page associated with a business goal
+client.track("visited-promo-page", evaluationContext);
```

Comment:
How can this be associated with a particular feature flag? Shouldn't flag key be included somewhere?

Reply:
There is no guarantee that a track invocation happens in a system that has made a feature flag evaluation. The user may have been the target and exposed to a specific value of a feature flag from some other system which impacts the user behaviour. 

The connection between "exposed to feature flag" and "track" can either be done by keeping track of feature flag evaluations, or as suggested in the diagram above, by using a hook to tell the analytics platform that a flag evaluation has happened. The important thing here is that we can provide the evaluation context with the targeting key which is how one would connect the feature flag resolve with the occurrence.

I'm not strongly opposing having a way to append optional metadata (for example flag evaluation results) in the track API, but I do think it introduces a somewhat unnecessary coupling that cannot be guaranteed in all instances.

Reply:
> There is no guarantee that a track invocation happens in a system that has made a feature flag evaluation. The user may have been the target and exposed to a specific value of a feature flag from some other system which impacts the user behaviour.
> 
> The connection between "exposed to feature flag" and "track" can either be done by keeping track of feature flag evaluations, or as suggested in the diagram above, by using a hook to tell the analytics platform that a flag evaluation has happened. The important thing here is that we can provide the evaluation context with the targeting key which is how one would connect the feature flag resolve with the occurrence.
> 
> I'm not strongly opposing having a way to append optional metadata (for example flag evaluation results) in the track API, but I do think it introduces a somewhat unnecessary coupling that cannot be guaranteed in all instances.

I couldn't have said it better.

Nothing is stopping anyone for adding some flag key in the additional data, but associating user actions with the flag evaluations is primarily the job of the backend flag or analytics platform(s). IMO Flag evaluations and are best thought of as separate transactions that are joined by a user session; And it may be that the flag impacting your business goals isn't even the one some developer though it was. In many ways I think tightly coupling flag evaluations to tracking the way @beeme1mr  is antithetical to the most robust form or experimentation... just looking at the data and querying it for associations, rather than baking some assumptions into the code.

Reply:
Thanks for the clarification.